state/protocol/protocol_state/kvstore.go

```diff
@@ -0,0 +1,108 @@
+type ViewBasedActivator[T any] struct {
+	Data           T
+	ActivationView uint64 // view at which data will be active
+}
+
+// VersionedEncodable defines the interface for a versioned key-value store independent
+// of the set of keys which are supported. This allows the storage layer to support
+// storing different key-value model versions within the same software version.
+type VersionedEncodable interface {
+	// VersionedEncode encodes the key-value store, returning the version separately
+	// from the encoded bytes.
+	// No errors are expected during normal operation.
+	VersionedEncode() (uint64, []byte, error)
+}
+
+// This file contains versioned read and read-write interfaces to the Protocol State's
+// key-value store and are used by the Protocol State Machine.
+//
+// When a key is added or removed, this requires a new protocol state version:
+//  - Create a new versioned model in ./kvstore/models.go (eg. modelv3 if latest model is modelv2)
+//  - Update the KVStoreReader and KVStoreAPI interfaces to include any new keys
+
+// KVStoreReader is the latest read-only interface to the Protocol State key-value store
+// at a particular block.
+// It is backward-compatible with all versioned model types defined in the models.go
+// for this software version.
+type KVStoreReader interface {
+	ID() flow.Identifier
+
+	// v0
+
+	VersionedEncodable
+
+	// GetProtocolStateVersion returns the current Protocol State version.
+	// The Protocol State version specifies the current version of the Protocol State,
+	// including the key-value store. Changes in the protocol state version
+	// correspond to changes in the set of key-value pairs which are supported,
+	// and which model is used for serialization.
+	// It can be updated by an UpdateKVStoreVersion service event.
+	GetProtocolStateVersion() uint64
+
+	// GetVersionUpgrade returns the upgrade version of protocol.
+	// VersionUpgrade is a view-based activator that specifies the version which has to be applied
+	// and the view at which it has to be applied. It may contain an old value if the upgrade has already applied.
+	// It can be updated by an flow.ProtocolStateVersionUpgrade service event.
+	GetVersionUpgrade() *ViewBasedActivator[uint64]
+
+	// v1
+
+	// GetInvalidEpochTransitionAttempted returns a flag indicating whether epoch
+	// fallback mode has been tentatively triggered on this fork.
+	// Errors:
+	//  - ErrKeyNotSupported if the key does not exist in the current model version.
+	GetInvalidEpochTransitionAttempted() (bool, error)
+}
+
+// KVStoreAPI is the latest read-writer interface to the Protocol State key-value store.
+// It is backward-compatible with all versioned model types defined in the models.go
+// for this software version.
+type KVStoreAPI interface {
+	KVStoreReader
+
+	// Clone returns a deep copy of the KVStoreAPI.
+	// This is used to create a new instance of the KVStoreAPI to avoid mutating the original.
+	Clone() KVStoreAPI
```

Comment:
Thanks for detailed explanation and argumentation, the more I read it, the more I tend to get rid of `Clone/Replicate` as I have implemented it in the first place. I feel like we are mixing two concepts, _mutating_ data and _creation_ process. 
I have tried to replicate the same structure as we have with epoch protocol state but it doesn't seem to work out. Meaning the state machine doesn't have enough context to provide a correct version to make a replication with correct version even if it had we are mixing creation pattern and behavior(factory + state machine). 

In my mind it works well where the `stateMutator` acts as a factory and creates a correct versioned `KVStoreMutator` which is then passed to the state machine. Similarly to what we have with the epoch state machine. This way we separate creation and mutation. This means we don't need `Replicate` or `Clone` as part of API and the state machine won't take a copy on it's own, instead it will rely on the instance that will be received via constructor from a component that has needed context to create it. 

Lets continue this discussion.

Reply:
### Summary of sync (Yurii, Jordan, Alex, March 12, 2024)

1. `Replicate(..)` or `Clone` must produce valid Snapshot, because in many blocks there are no service events
2. from 1. it follows that we don't need builder pattern 
3. extending `stateMutator` to be the orchestrator of all state machines (currently only orchestrates Epoch State machine) 
    * `stateMutator` will call `Replicate`
    * state machines are applied, they mutate the combined state in-place 
    * we `Build` method in  `stateMutator`  where we could add sanity checks

We want to move the entire Epoch data into the KV-store, but that is beyond the current deliverables.